### PR TITLE
EZEE-2915: Implement minimal test coverage for Richtext block in Page Builder

### PR DESCRIPTION
**Pull request by katarzynazawada**
| Question      | Answer
| ------------- | ---
| Tickets       |  https://jira.ez.no/browse/EZEE-2915
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Tests pass?   | yes
| Doc needed?   | no
| License       | [GPL-2.0](https://github.com/ezsystems/ezplatform-admin-ui/blob/master/LICENSE)

#### Checklist:
- [x] Coding standards (`$ composer fix-cs`)
- [x] Ready for Code Review

### CODE REVIEW

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 1]
```diff
@@ -8,43 +8,44 @@
...
         parent::__construct($context, $locator, $label);
         $this->fields['fieldInput'] = '.ez-data-source__richtext';
         $this->fields['textarea'] = $this->fields['fieldContainer'] . ' textarea';
+        $this->fields['embedInlineButton'] = '.ez-btn-ae--embed-inline';
+        $this->fields['embedButton'] = '.ez-btn-ae--embed';
+        $this->fields['addButton'] = '.ae-button-add';
+        $this->fields['embedTitle'] = '.cke_widget_ezembed .ez-embed-content__title';
+        $this->fields['embedInlineTitle'] = '.cke_widget_ezembedinline .ez-embed-content__title';
+        $this->fields['unorderedListButton'] = '.ez-btn-ae--unordered-list';
+        $this->fields['unorderedList'] = '.ez-data-source__richtext ul li';
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]: `unorderedListElement`?

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 2]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
             'Field has wrong value'
         );
     }
+
+    public function openElementsToolbar()
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]:
```suggestion
    public function openElementsToolbar(): void
```

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 3]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+    {
+        $this->context->findElement($this->fields['addButton'])->click();
+    }
+
+    public function changeStyle(string $style)
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]:
```suggestion
    public function changeStyle(string $style): void
```

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 4]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+    private function getFieldInput(): NodeElement
+    {
+        $fieldInput = $this->context->findElement($this->fields['fieldInput']);
+        Assert::assertNotNull($fieldInput, sprintf('Input for field %s not found.', $this->label));
+        $this->richtextId = $fieldInput->getAttribute('id');
+
+        return $fieldInput;
+    }
+
+    public function clickOnRichTextEditor()
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]: Is it used? I can't find any usages in this PR or in https://github.com/ezsystems/ezplatform-page-builder/pull/421/files (but I might be missing something)

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 5]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+    {
+        $fieldInput = $this->getFieldInput();
+        $fieldInput->click();
+    }
+
+    public function addUnorderedList(array $listElements)
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]:
```suggestion
    public function addUnorderedList(array $listElements): void
```

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 6]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+        $actualListElements = $this->context->findAllElements($this->fields['unorderedList']);
+        $listElementsText = [];
+        foreach ($actualListElements as $actualListElement) {
+            $listElementsText[] = $actualListElement->getText();
+        }
+
+        Assert::assertEquals($listElements, $listElementsText);
+    }
+
+    public function clickEmbedInlineButton()
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]:
```suggestion
    public function clickEmbedInlineButton(): void
```

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 7]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+    {
+        $this->context->findElement($this->fields['embedInlineButton'])->click();
+    }
+
+    public function clickEmbedButton()
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]:
```suggestion
    public function clickEmbedButton(): void
```

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 8]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+    {
+        $this->context->findElement($this->fields['embedButton'])->click();
+    }
+
+    public function containsEmbedInlineItem($itemName)
+    {
+        Assert::assertContains($itemName, $this->context->findElement($this->fields['embedInlineTitle'])->getText());
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]: This is different from the convention that we're using, that is:
- if the function asserts something then it should be named `verify....` (verifyContainsEmbedInlineItem in this case).
- if the function name is a question (`contains Embed Inline item?`) then it should return a boolean value (`return $itemName === $this->context->findElement($this->fields['embedInlineTitle'])->getText()`) and the assertion should happen here: https://github.com/ezsystems/ezplatform-page-builder/pull/421/files#diff-ab43e5513485e9364a60c01a53e26debR86

I'd go with a boolean here and put the assertion in PB. This comment also applies to `containsEmbedItem`.

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 9]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+    }
+
+    public function containsEmbedItem($itemName)
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]:
```suggestion
    public function containsEmbedItem($itemName): void
```

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 10]
```diff
@@ -57,4 +58,103 @@ public function verifyValueInItemView(array $values): void
...
+    {
+        Assert::assertContains($itemName, $this->context->findElement($this->fields['embedTitle'])->getText());
+    }
+
+    public function moveElement($direction)
```
Comment (mnocon) [inline src/lib/Behat/PageElement/Fields/RichText.php]:
```suggestion
    public function moveElement($direction): void
```

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 11]
```diff
@@ -57,4 +58,97 @@ public function verifyValueInItemView(array $values): void
...
+    }
+
+    public function changeStyle(string $style): void
+    {
+        if (!in_array($style, self::ALLOWED_STYLE_OPTIONS)) {
+            throw new Exception(sprintf('Unsupported style: %s', $style));
+        }
+
+        $this->context->findElement($this->fields['styleDropdown'])->click();
+        $this->context->findElement(sprintf('.ae-listbox li %s', $style))->click();
```
Comment (m-tyrala) [inline src/lib/Behat/PageElement/Fields/RichText.php]: I know it's a short one, but I'd prefer to add it to the fields table.

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 12]
```diff
@@ -57,4 +58,97 @@ public function verifyValueInItemView(array $values): void
...
+    }
+
+    public function insertLine($value, $style = ''): void
+    {
+        $this->getFieldInput();
+        $this->context->getSession()->getDriver()->executeScript(sprintf($this->insertAlloyEditorValueScript, $this->richtextId, $value));
+
+        if ($style !== '') {
+            $this->changeStyle($style);
+            Assert::assertContains(sprintf('%s%s</%s>', $value, '<br>', $style), $this->context->findElement(sprintf('.ez-data-source__richtext %s', $style))->getOuterHtml());
```
Comment (m-tyrala) [inline src/lib/Behat/PageElement/Fields/RichText.php]: I'd appreciate if you'd replace `.ez-data-source__richtext` with the second %s, and add the `$this->fields['fieldInput']` to the `sprintf` parameters.

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 13]
```diff
@@ -57,4 +58,97 @@ public function verifyValueInItemView(array $values): void
...
+    }
+
+    public function clickEmbedButton(): void
+    {
+        $this->context->findElement($this->fields['embedButton'])->click();
+    }
+
+    public function containsEmbedInlineItem($itemName): bool
+    {
+        return $itemName === $this->context->findElement($this->fields['embedInlineTitle'])->getText();
```
Comment (m-tyrala) [inline src/lib/Behat/PageElement/Fields/RichText.php]: It's rather `equalsEmbedInlineItem` than `contains(...)` (and the same thing below).

File: src/lib/Behat/PageElement/Fields/RichText.php [hunk 14]
```diff
@@ -57,4 +58,97 @@ public function verifyValueInItemView(array $values): void
...
+        return $itemName === $this->context->findElement($this->fields['embedTitle'])->getText();
+    }
+
+    public function moveElement($direction): void
+    {
+        if (!in_array($direction, self::ALLOWED_MOVE_OPTIONS)) {
+            throw new Exception(sprintf('Unsupported direction: %s', $direction));
+        }
+
+        $this->context->findElement(sprintf('.ez-btn-ae--move-%s', $direction))->click();
```
Comment (m-tyrala) [inline src/lib/Behat/PageElement/Fields/RichText.php]: Same as above - don't hardcode locators please, add it to the fields table.

**Pull request merged**